typed-racket-lib/typed-racket/static-contracts/combinators/dep-function.rkt

```diff
@@ -0,0 +1,79 @@
+#lang racket/base
+
+;; Static contract for dependent ->.
+
+(require "../../utils/utils.rkt"
+         "../structures.rkt" "../constraints.rkt"
+         racket/match
+         racket/list
+         (contract-req))
+
+(define depth 0)
+
+(struct dep-function/sc static-contract (typed-side?
+                                         ids
+                                         dom
+                                         dom-deps
+                                         rng
+                                         rng-deps)
+  #:transparent
+  #:property prop:combinator-name "dep->/sc"
+  #:methods gen:sc
+  [(define (sc->contract v rec)
+     (match v
+       [(dep-function/sc typed-side? ids dom/scs deps rng/scs rng-deps)
+        (with-syntax ([(id ...) ids]
+                      [(c ...) (cond
+                                 [typed-side?
+                                  (for/list ([d/sc (in-list dom/scs)]
+                                             [dep-ids (in-list deps)])
+                                    (cond
+                                      [(not (null? dep-ids))
+                                       (parameterize ([static-contract-may-contain-free-ids? #t])
+                                         (rec d/sc))]
+                                      [else (rec d/sc)]))]
+                                   [else (make-list (length dom/scs) #'any/c)])]
+                      [(dep ...) deps]
+                      [(r-deps ...) rng-deps])
+          #`(->i ([id dep c] ...)
+                 #,(cond
+                     [typed-side? #'any]
```

Comment:
I'm worried about this because of #598 .
Is `typed-side?` definitely `#false` when the return type is a higher-order value?

Reply:
Ok, I think you should definitely delete this line.

Instead, let the optimizer reduce the range to `any/sc` when that makes sense, and add a case to [`static-contracts/optimize.rkt`](https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/static-contracts/optimize.rkt#L115) to let the function here (I mean `sc->contract`) know when it should replace the range with `any`.

Reply:
I dropped the `typed-side?` optimizations -- glancing at all that code, it's not obvious to me what optimizations for dependent functions are worth adding to optimize.rkt... so perhaps I'll leave that for another day ;-)

Reply:
Okay, I thought about it a little more and came up w/ a simple case we can safely replace the range with `any` for after the optimization pass (as you suggested):

```racket
[(and typed-side? (andmap any/sc? rng-deps)) #'any]
```